File: ResourceManager/ViewModels/Repositories/RepositoryContentViewModel.vb
Imports VBNetCore.Mvvm.Abstractions
Imports VBNetCore.Mvvm.Commands
Imports VBNetCore.Mvvm.Controls
Imports ResourceManager.Configuration
Imports ResourceManager.Helpers
Imports ResourceManager.Models
Imports ResourceManager.Services.Abstractions
Imports System.Diagnostics
Imports System.IO
Imports System.Linq
Imports System.Threading.Tasks
Imports System.Windows.Input
Imports Microsoft.Extensions.DependencyInjection

Namespace ResourceManager.ViewModels
    Public Class RepositoryContentViewModel
        Inherits PageContentWithListViewModel(Of Repository)
        Private ReadOnly _dataServiceManager As IDataServiceManager
        Public Sub New(ByVal dataServiceManager As IDataServiceManager)
            MyBase.New(dataServiceManager.RepositoryDataService)
            _dataServiceManager = dataServiceManager
            MenuTitle = "Repository"
            MenuIcon = AssetManager.Instance.GetImage("Repository.png")
            Title = "Repository Manager"
            PageColor = RepositoryPageColor

            _edit_permission_name = ACTION_EDIT_REPOSITORY
            _add_permission_name = ACTION_ADD_REPOSITORY
            _delete_permission_name = ACTION_DEL_REPOSITORY

            TagsCommand = New AsyncRelayCommand(Of Repository)(AddressOf OnTagsAsync)
            ExploreCommand = New RelayCommand(Of Repository)(AddressOf OnExplore)

            MyBase.OnRefresh()
        End Sub

        Private Sub OnExplore(ByVal repo As Repository)
            Dim filename = repo.FilePath

            If Not File.Exists(filename) Then
                ShowPopupMessage("File '" & filename & "' doesn't exists.", Application.Settings.AppName, PopupButton.OK, PopupImage.[Error])
                Return
            End If
            Dim process = New Process()
            Dim startInfo = New ProcessStartInfo()
            startInfo.FileName = "explorer.exe"

            ' sqlite needs to have two slashes instead of one for the path it uses inside sqlite3
            startInfo.Arguments = $"/select,""{filename}"""
            startInfo.CreateNoWindow = True
            startInfo.UseShellExecute = False
            process.StartInfo = startInfo
            process.Start()
            process.WaitForExit()
            process.Close()
        End Sub

        Private Async Function OnTagsAsync(ByVal repo As Repository) As Task
            Dim vm = Application.ServiceProvider.GetRequiredService(Of TagsPopupViewModel)()
            vm.Caption = String.Format("Manage tag for {0}", repo.GetCaption())
            Dim tags = Await _dataServiceManager.TagDataService.GetTagObjectForRepositoryID(repo.ID)
            vm.Tags = tags.ToList()
            Dim popup = ShowPopup(vm)
            If popup.Result = PopupResult.OK Then
                Await _dataServiceManager.RepositoryDataService.UpdateTags(repo, vm.Tags)
                MyBase.OnRefresh()
            End If
        End Function


        Protected Overrides Sub OnEdit(ByVal repo As Repository)
            Dim vm = Application.ServiceProvider.GetRequiredService(Of RepositoryPopupViewModel)()
            vm.Caption = "Edit Repository"
            vm.SetRepo(repo)
            Dim popup = ShowPopup(vm)
            If popup.Result = PopupResult.OK Then
                MyBase.OnRefresh()
            End If
        End Sub

        Protected Overrides Sub OnAdd()
            Dim vm = Application.ServiceProvider.GetRequiredService(Of RepositoryPopupViewModel)()
            vm.Caption = "Add Repository"
            Dim popup = ShowPopup(vm)
            If popup.Result = PopupResult.OK Then
                MyBase.OnRefresh()
            End If
        End Sub
        Protected Overrides Sub OnFilter()
            MyBase.Filter(FilterColumns.Repository, "Filter Repository")
        End Sub

        Public ReadOnly Property TagsCommand As ICommand
        Public ReadOnly Property ExploreCommand As ICommand
    End Class
End Namespace

File: ResourceManager/ViewModels/Repositories/RepositoryContentViewModel.cs
using Microsoft.Extensions.DependencyInjection;
using NetCore.Mvvm.Abstractions;
using NetCore.Mvvm.Commands;
using NetCore.Mvvm.Controls;
using ResourceManager.Configuration;
using ResourceManager.Helpers;
using ResourceManager.Models;
using ResourceManager.Services;
using ResourceManager.Services.Abstractions;
using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Input;

namespace ResourceManager.ViewModels
{
    public class RepositoryContentViewModel : PageContentWithListViewModel<Repository>
    {
        private readonly IDataServiceManager _dataServiceManager;
        public RepositoryContentViewModel(IDataServiceManager dataServiceManager)
            : base(dataServiceManager.RepositoryDataService)
        {
            _dataServiceManager = dataServiceManager;
            MenuTitle = "Repository";
            MenuIcon = AssetManager.Instance.GetImage("Repository.png");
            Title = "Repository Manager";
            PageColor = Constants.RepositoryPageColor;

            _edit_permission_name = Constants.ACTION_EDIT_REPOSITORY;
            _add_permission_name = Constants.ACTION_ADD_REPOSITORY;
            _delete_permission_name = Constants.ACTION_DEL_REPOSITORY;

            TagsCommand = new AsyncRelayCommand<Repository>(OnTagsAsync);
            ExploreCommand = new RelayCommand<Repository>(OnExplore);

            OnRefresh();
        }

        private void OnExplore(Repository repo)
        {
            string filename = repo.FilePath;

            if (!File.Exists(filename))
            {
                ShowPopupMessage("File '" + filename + "' doesn't exists.", App.Settings.AppName, PopupButton.OK, PopupImage.Error);
                return;
            }
            var process = new Process();
            var startInfo = new ProcessStartInfo();
            startInfo.FileName = "explorer.exe";

            // sqlite needs to have two slashes instead of one for the path it uses inside sqlite3
            startInfo.Arguments = $@"/select,""{filename}""";
            startInfo.CreateNoWindow = true;
            startInfo.UseShellExecute = false;
            process.StartInfo = startInfo;
            process.Start();
            process.WaitForExit();
            process.Close();
        }

        private async Task OnTagsAsync(Repository repo)
        {
            var vm = App.ServiceProvider.GetRequiredService<TagsPopupViewModel>();
            vm.Caption = string.Format("Manage tag for {0}", repo.GetCaption());
            var tags = await _dataServiceManager.TagDataService.GetTagObjectForRepositoryID(repo.ID);
            vm.Tags = tags.ToList();
            var popup = ShowPopup(vm);
            if (popup?.Result == PopupResult.OK)
            {
                await _dataServiceManager.RepositoryDataService.UpdateTags(repo, vm.Tags);
                OnRefresh();
            }
        }


        protected override void OnEdit(Repository repo)
        {
            var vm = App.ServiceProvider.GetRequiredService<RepositoryPopupViewModel>();
            vm.Caption = "Edit Repository";
            vm.SetRepo(repo);
            var popup = ShowPopup(vm);
            if (popup?.Result == PopupResult.OK)
            {
                OnRefresh();
            }
        }

        protected override void OnAdd()
        {
            var vm = App.ServiceProvider.GetRequiredService<RepositoryPopupViewModel>();
            vm.Caption = "Add Repository";
            var popup = ShowPopup(vm);
            if (popup?.Result == PopupResult.OK)
            {
                OnRefresh();
            }
        }
        protected override void OnFilter()
        {
            Filter(FilterColumns.Repository, "Filter Repository");
        }

        public ICommand TagsCommand { get; }
        public ICommand ExploreCommand { get; }
    }
}
